lib/THC/THCTensorMode.cuh

```diff
@@ -0,0 +1,282 @@
+#ifndef THC_TENSOR_MODE_CUH
+#define THC_TENSOR_MODE_CUH
+
+#include "THCNumerics.cuh"
+#include "THCSortUtils.cuh"
+
+struct ThrustHalfLess
+{
+  __host__ __device__ inline bool operator()(const half& lhs, const half& rhs) {
+    return THCNumerics<half>::lt(lhs, rhs);
+  }
+};
+
+struct ThrustHalfNotEqualTo
+{
+  __host__ __device__ inline bool operator()(const half& lhs, const half& rhs) {
+    return THCNumerics<half>::ne(lhs, rhs);
+  }
+};
+
+struct ThrustHalfEqualTo
+{
+  __host__ __device__ inline bool operator()(const half& lhs, const half& rhs) {
+    return THCNumerics<half>::eq(lhs, rhs);
+  }
+};
+
+struct ThrustHalfEqualToPredicate
+{
+  ThrustHalfEqualToPredicate(half val): val_(val) {}
+  __host__ __device__ inline bool operator()(half x) {
+    return THCNumerics<half>::eq(val_, x);
+  }
+
+  half val_;
+};
+
+template <typename T>
+struct BinaryAddOp {
+  __host__ __device__ inline T operator()(const T a, const T b) {
+    return THCNumerics<T>::add(a, b);
+  }
+};
+
+template <>
+struct BinaryAddOp<unsigned int> {
+  __host__ __device__ inline unsigned int operator()(const unsigned int a, const unsigned int b) {
+    return a + b;
+  }
+};
+
+template <typename T, class BinaryOp, int Power2ScanSize>
+__device__ void segmentedInclusivePrefixScan(T *smem, bool *bmem, BinaryOp binop) {
```

Comment:
move these segmented scans into the THCReduceApplyUtils so they can be usable elsewhere (a prefix scan is a kind of reduction, it gives all of the partial reductions of a sequence)